### PR TITLE
[Bounty] mrat

### PR DESCRIPTION
**Pull request by ynot01**
# Github documenting your Pull Request

Request Mentor Rat button
polls all mentors that are ghosts
selects from those who said yes
makes imaginary friend who is a rat and can leave when they choose
their name is their ckey
https://forums.yogstation.net/threads/15-mentor-mouse-aka-mrat-ss13-server.22889/

image
image
image
image
image

## user side
image
image
image

## mentor side
image
image
image
image

# Wiki Documentation

Request Mentor Rat button in Mentor tab

# Changelog

:cl:  
rscadd: mrat  
/:cl:

### CODE REVIEW

**Comment by Reed0506** [comment]
mrat mrat mrat mrat mrat mrat mrat

File: code/__HELPERS/game.dm [hunk 1]
```diff
@@ -484,6 +484,31 @@
...
+  * * ignore_category: Define, ignore_category: People with this category(defined in poll_ignore.dm) turned off dont get the message
+  * * flashwindow: Bool, Flash their window to grab their attention
+  */
+/proc/pollMentorGhostCandidates(Question, jobbanType, datum/game_mode/gametypeCheck, be_special_flag = 0, poll_time = 300, ignore_category = null, flashwindow = TRUE)
+	var/list/candidates = list()
+	if(!(GLOB.ghost_role_flags & GHOSTROLE_STATION_SENTIENCE))
+		return candidates
+
+	for(var/mob/dead/observer/G in GLOB.player_list)
+		if(!is_mentor(G))
```
Comment (monster860) [inline code/__HELPERS/game.dm]: i think this does the oppoite of what you want

**Review by monster860 — DISMISSED** [review]

File: code/__HELPERS/game.dm (hunk 1 above)
Comment (ynot01) [reply code/__HELPERS/game.dm]: oh shit woops

**Comment by N3D6** [comment]
best pr

File: code/datums/brain_damage/mrat.dm [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/datum/brain_trauma/special/imaginary_friend/mrat/get_ghost()
+	set waitfor = FALSE
+	var/list/mob/dead/observer/candidates = pollMentorCandidatesForMob("Do you want to play as [owner]'s mentor rat?", ROLE_PAI, null, null, 75, friend, POLL_IGNORE_IMAGINARYFRIEND)
+	if(LAZYLEN(candidates))
+		var/mob/dead/observer/C = pick(candidates)
+		friend.key = C.key
+		friend_initialized = TRUE
+		to_chat(owner, "<span class='notice'>You have acquired the mentor rat known as [friend.key], ask them any question you like.</span>")
+	else
```
Comment (Bibby0110) [inline code/datums/brain_damage/mrat.dm:26]: Probably need some kind of response if no mentors accept

**Comment by Bibby0110** [comment]
Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
Name it something that actually conveys what it is about

File: code/datums/brain_damage/mrat.dm [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+/mob/camera/imaginary_friend/mrat/greet()
+		to_chat(src, "<span class='notice'><b>You are the mentor rat of [owner]!</b></span>")
+		to_chat(src, "<span class='notice'>Do not give [owner] any OOC information from your time as a ghost.</span>")
+		to_chat(src, "<span class='notice'>Your job is to answer [owner]'s question(s) and you are given this form to assist in that.</span>")
+		to_chat(src, "<span class='notice'>Don't be stupid with this or you will face the consequences.</span>")
+
+/mob/camera/imaginary_friend/mrat/Initialize(mapload, _trauma)
+	. = ..()
+
+	trauma = _trauma
```
Comment (Bibby0110) [inline code/datums/brain_damage/mrat.dm]: Everything in this proc expect the stuff about the leave action is already done in the parent proc

File: yogstation/code/modules/mentor/mentor_verbs.dm [hunk 4]
```diff
@@ -70,6 +70,17 @@ GLOBAL_PROTECT(mentor_verbs)
 		msg += "<span class='info'>Mentorhelps are also seen by admins. If no mentors are available in game adminhelp instead and an admin will see it and respond.</span>"
 	to_chat(src, msg, confidential=TRUE)
 
+/client/verb/mrat()
+	set name = "Request Mentor Rat"
+	set category = "Mentor"
+
+	var/mob/living/carbon/human/C = src.mob
```
Comment (Bibby0110) [inline yogstation/code/modules/mentor/mentor_verbs.dm]: Probably want to verify that they’re a carbon

File: code/datums/brain_damage/mrat.dm [hunk 5]
```diff
@@ -0,0 +1,63 @@
+/datum/brain_trauma/special/imaginary_friend/mrat
```
Comment (Bibby0110) [inline code/datums/brain_damage/mrat.dm:1]: make_friend proc is missing?

**Review by Bibby0110 — COMMENTED** [review]
Can you verify that imaginary friends can point?

Might also want to check if runechat works, might be a little confusing for a new player if that’s that’s what they’re used to

File: code/datums/brain_damage/mrat.dm (hunk 2 above)
Comment (ynot01) [reply code/datums/brain_damage/mrat.dm:26]: if no mentors accept it triggers lose_text
`lose_text = "<span class='warning'>Mentor rat not found. Either your mentor rat left you or you didn't find one.</span>"`

File: code/datums/brain_damage/mrat.dm (hunk 3 above)
Comment (ynot01) [reply code/datums/brain_damage/mrat.dm]: fixed

File: yogstation/code/modules/mentor/mentor_verbs.dm (hunk 4 above)
Comment (ynot01) [reply yogstation/code/modules/mentor/mentor_verbs.dm]: will do

File: code/datums/brain_damage/mrat.dm (hunk 5 above)
Comment (ynot01) [reply code/datums/brain_damage/mrat.dm:1]: not anymore

File: yogstation/code/modules/mentor/mentor_verbs.dm (hunk 4 above)
Comment (ynot01) [reply yogstation/code/modules/mentor/mentor_verbs.dm]: done

**Comment by ynot01** [comment]
> Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
> Name it something that actually conveys what it is about

bounty says mrat
if you have any ideas im welcome to them

File: code/datums/brain_damage/mrat.dm (hunk 2 above)
Comment (Bibby0110) [reply code/datums/brain_damage/mrat.dm:26]: Ah fair, didn’t see that

**Comment by Bibby0110** [comment]
> > Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
> > Name it something that actually conveys what it is about
> 
> bounty says mrat
> if you have any ideas im welcome to them

Something boring like, request mentor assistance. Maybe make a prompt that tells the person what this is and has a yes/no option, I feel like this might get clicked a lot just to figure out what it does

**Comment by ynot01** [comment]
> > > Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
> > > Name it something that actually conveys what it is about
> > 
> > 
> > bounty says mrat
> > if you have any ideas im welcome to them
> 
> Something boring like, request mentor assistance. Maybe make a prompt that tells the person what this is and has a yes/no option, I feel like this might get clicked a lot just to figure out what it does

i could do this

**Comment by ynot01** [comment]
> > > Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
> > > Name it something that actually conveys what it is about
> > 
> > 
> > bounty says mrat
> > if you have any ideas im welcome to them
> 
> Something boring like, request mentor assistance. Maybe make a prompt that tells the person what this is and has a yes/no option, I feel like this might get clicked a lot just to figure out what it does

**Comment by ynot01** [comment]
> Can you verify that imaginary friends can point?
> 
> Might also want to check if runechat works, might be a little confusing for a new player if that’s that’s what they’re used to

Unfortunately, runechat does not work on the rat and my efforts have yielded nothing in face of it.
As for pointing, I have gotten the chat to show but the arrow does not.

As workarounds, I have added a squeak whenever the mrat talks and as for pointing, you can still point but it will not show an arrow. Best to point using your 'physical' body.

**Comment by Bibby0110** [comment]
> > > > Mentor rat is a really horrible and non-descriptive name for anyone new trying to use this.
> > > > Name it something that actually conveys what it is about
> > > 
> > > 
> > > bounty says mrat
> > > if you have any ideas im welcome to them
> > 
> > 
> > Something boring like, request mentor assistance. Maybe make a prompt that tells the person what this is and has a yes/no option, I feel like this might get clicked a lot just to figure out what it does
> 
> 

Very nice!

**Comment by Bibby0110** [comment]
> > Can you verify that imaginary friends can point?
> > Might also want to check if runechat works, might be a little confusing for a new player if that’s that’s what they’re used to
> 
> Unfortunately, runechat does not work on the rat and my efforts have yielded nothing in face of it.
> As for pointing, I have gotten the chat to show but the arrow does not.
> 
> As workarounds, I have added a squeak whenever the mrat talks and as for pointing, you can still point but it will not show an arrow. Best to point using your 'physical' body.

I’ll look into it

**Comment by alexkar598** [comment]
Okay so I love the idea but I have to downvote this for the reason I posted on the forums 6 months ago
> I get the meme and haha funny but since this is more aimed at new players who lack the basic knowledge required to bootstrap learning new things based on text alone, a rat has too much potential to be confusing and leaving people wondering what the fuck is happening. If this is to be, it really should be clearly marked as an OOC component and not a potentially IC one. Not everyone is into mentor "culture"
 
You see rats wondering around the station in game, you see player controlled rats wondering around the station in game. SS13 is a very confusing game with an insane learning curve especially early on, making players understand that the rat is not only not a real rat but will offer them OOC guidance (which blurs the lines between IC and OOC, causing the player issues with rules later on) will result into an unintuitive learning experience to satisfy a "haha mrat" meme.

Again, the idea is great but the execution is terrible, I believe the best way to handle this would be with a modified AI camera sprite which can speak such as this: amentor

**Comment by alexkar598** [comment]
After arguing on discord it seems noone cares what I think so hahaha funny mrat i love this pr

**Comment by maxion12345** [comment]
> After arguing on discord it seems noone cares what I think so hahaha funny mrat i love this pr

Good, good! Embrace the mrat

Speaking of, can they pet the mrat? 
Also "request mentor assistance" is meh. What about "Summon mentor assistant"?

**Comment by Smudged** [comment]
> After arguing on discord it seems noone cares what I think so hahaha funny mrat i love this pr

Well I will say these are valid critisisms, maybe making the rat visually distinct would help relax confusion, it's something important to look at and I hope people take your criticism into perspective.

**Comment by Darkstick** [comment]
does it walk fast enough to keep up with the player?

**Comment by ynot01** [comment]
its green now

**Comment by ynot01** [comment]
> does it walk fast enough to keep up with the player?

it can go through walls and moves at ghost speeds

**Comment by ynot01** [comment]
lots of new stuff see top post for screenshots

**Comment by JamieD1** [comment]
Make it the mentor colour!

**Review by JamieD1 — APPROVED** [review]

**Comment by Moja10** [comment]
My holopad suggestion was not implemented, Literally 1984

**Pull request merged**